### PR TITLE
Update black to 21.4b0, run pre-commit checks, correctly raise errors

### PR DESCRIPTION
**Pull request by Zeitsperre**
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #xyz
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
- [x] `bumpversion (major / minor / patch)` has been called on this branch
- [ ] Tags have been pushed (`git push --tags`)

* **What kind of change does this PR introduce?** 

Updates pre-commit hooks, most notably Black. Black now reformats docstrings as needed and has implemented some new conventions. Very little changes for us. For more information, see: https://github.com/psf/black/releases/tag/21.4b0

* **Does this PR introduce a breaking change?** 

Nope.

* **Other information**:

Updates should be automatic if `pre-commit is installed and initialized.

### CODE REVIEW

**Review by aulemahal — APPROVED** [review]

**Comment by Zeitsperre** [comment]
@aulemahal Tahnks for review. I'm adding a bit more to this, mostly docstrings. Will merge when ready.

File: xclim/core/calendar.py [hunk 1]
```diff
@@ -337,7 +340,7 @@ def _convert_datetime(
     datetime: Union[pydt.datetime, cftime.datetime],
     new_doy: Optional[Union[float, int]] = None,
     calendar: str = "default",
-):
+) -> Union[cftime.datetime, pd.DatetimeIndex, np.int]:
```
Comment (aulemahal) [inline xclim/core/calendar.py:343]:
```suggestion
) -> Union[cftime.datetime, pydt.datetime, np.int]:
```
Would be more exact.

File: xclim/core/calendar.py [hunk 2]
```diff
@@ -353,7 +356,7 @@ def _convert_datetime(
 
     Returns
     -------
-    Union[cftime.datetime, pydt.datetime, np.nan]
+    Union[cftime.datetime, pandas.DatetimeIndex, np.nan]
```
Comment (aulemahal) [inline xclim/core/calendar.py:359]:
```suggestion
    Union[cftime.datetime, datetime.datetime, np.nan]
```

File: xclim/core/calendar.py [hunk 3]
```diff
@@ -379,7 +382,9 @@ def _convert_datetime(
         return np.nan
 
 
-def ensure_cftime_array(time: Sequence):
+def ensure_cftime_array(
+    time: Sequence,
+) -> Union[cftime.DatetimeGregorian, numpy.ndarray]:
```
Comment (aulemahal) [inline xclim/core/calendar.py:387]:
```suggestion
) -> Union[CFTimeIndex, numpy.ndarray]:
```
Not clear from the code itself, but I think those are the only two possibilities!

**Review by aulemahal — COMMENTED** [review]
Thanks for all this, it is cleaner like this.

**Comment by Zeitsperre** [comment]
`scipy.spatial.cKDTree().query()` does not accept the `workers` argument in `scipy<1.16` which is causing build errors for my existing xclim env (https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.query.html#scipy.spatial.cKDTree.query).

Are there any reasons to justify not raising the minimum `scipy` version?

**Comment by aulemahal** [comment]
Well, sadly there's #684... Should we change the kwarg name according to the installed version?

**Comment by Zeitsperre** [comment]
If there's no way around it, that might be the quickest solution outside porting L-moments3 ourselves.

**Comment by Zeitsperre** [comment]
Found a quicker solution. NN and KLDIV tests are now explicitly skipped for `scipy<1.6`.

**Pull request merged**